subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/checkstyle/CheckstylePluginToolchainsIntegrationTest.groovy

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins.quality.checkstyle
+
+import org.gradle.api.specs.Spec
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.internal.jvm.Jvm
+import org.gradle.internal.jvm.inspection.JvmInstallationMetadata
+import org.hamcrest.Matcher
+import org.junit.Assume
+
+import static org.gradle.util.Matchers.containsLine
+import static org.hamcrest.CoreMatchers.containsString
+
+class CheckstylePluginToolchainsIntegrationTest extends AbstractIntegrationSpec {
+
+    def "uses jdk from toolchains"() {
+        given:
+        writeDummyConfig()
+        file('src/main/java/Dummy.java') << "class Dummy {}"
+        def jdk = setupExecutorForToolchains {
+            it.languageVersion > Jvm.current().javaVersion
+        }
```

Comment:
I think what might make a little more sense here is to use `AvailableJavaHomes.getDifferentVersion` instead of an arbitrary filter. I think you have the right pieces because we want tests that show:

- You can analyze good and bad code when the toolchain JDK is _not_ the same as the current JDK
- We default to using the current JDK when nothing is configured
- We use a different JDK when it's explicitly set on the task
- We use a different JDK when it's explicitly set in the Java plugin